Sources/App/Core/SearchLogger.swift

```diff
@@ -0,0 +1,57 @@
+// Copyright Dave Verwer, Sven A. Schmidt, and other contributors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import Foundation
+
+struct SearchQuery: Hashable, Codable {
+    let searchID: UUID
+    let query: String
+}
+
+struct SearchResultFragment: Hashable, Codable {
+    let searchID: UUID
+    let result: Search.Result?
+
+    enum CodingKeys: String, CodingKey {
+        case searchID = "id"
+        case result = "r"
+    }
+}
+
+enum SearchLogger {
+    static func log(query: String, results: [Search.Result]) {
+        let uniqueSearchID = UUID()
+        let jsonEncoder = JSONEncoder()
+        jsonEncoder.dateEncodingStrategy = .iso8601
+        jsonEncoder.outputFormatting = .sortedKeys
+
+        let baseQuery = SearchQuery(searchID: uniqueSearchID, query: query)
+        do {
+            let stringdata = try jsonEncoder.encode(baseQuery)
+            AppEnvironment.logger.info("search: \(stringdata)")
+        } catch {
+            AppEnvironment.logger.warning("unable to encode search query: \(error)")
+        }
+
+        for result in results {
+            let fragment = SearchResultFragment(searchID: uniqueSearchID, result: result)
+            do {
+                let stringdata = try jsonEncoder.encode(fragment)
+                AppEnvironment.logger.info("searchresult: \(stringdata)")
+            } catch {
+                AppEnvironment.logger.warning("unable to encode search fragment: \(error)")
+            }
```

Comment:
So I've run this locally with your branch and I'm getting the following when I log this:

```
[ INFO ] GET /search [component: server, request-id: 5AB91460-7BB6-4876-9106-9BC04769ADF1]
[ INFO ] search: 65 bytes [component: server]
[ INFO ] searchresult: 90 bytes [component: server]
[ INFO ] searchresult: 91 bytes [component: server]
[ INFO ] searchresult: 94 bytes [component: server]
[ INFO ] searchresult: 96 bytes [component: server]
[ INFO ] searchresult: 422 bytes [component: server]
[ INFO ] searchresult: 436 bytes [component: server]
[ INFO ] searchresult: 416 bytes [component: server]
[ INFO ] searchresult: 437 bytes [component: server]
[ INFO ] searchresult: 451 bytes [component: server]
[ INFO ] searchresult: 407 bytes [component: server]
```

which is expected I think, because the return type is `Data`. We'd need to wrap that with `String(decoding:as:)` in lines 41 and 51:

```swift
            let json = String(decoding: try jsonEncoder.encode(baseQuery), as: UTF8.self)
            AppEnvironment.logger.info("search: \(json)")
```

That then gives us:

```
[ INFO ] search: {"query":"foo","searchID":"DD410B89-DF59-437F-9D41-65C06ED39E7E"} [component: server]
[ INFO ] searchresult: {"id":"DD410B89-DF59-437F-9D41-65C06ED39E7E","r":{"author":{"_0":{"name":"adamfootdev"}}}} [component: server]
[ INFO ] searchresult: {"id":"DD410B89-DF59-437F-9D41-65C06ED39E7E","r":{"author":{"_0":{"name":"onefootprint"}}}} [component: server]
[ INFO ] searchresult: {"id":"DD410B89-DF59-437F-9D41-65C06ED39E7E","r":{"author":{"_0":{"name":"autoreleasefool"}}}} [component: server]
[ INFO ] searchresult: {"id":"DD410B89-DF59-437F-9D41-65C06ED39E7E","r":{"keyword":{"_0":{"keyword":"food-ordering"}}}} [component: server]
[ INFO ] searchresult: {"id":"DD410B89-DF59-437F-9D41-65C06ED39E7E","r":{"package":{"_0":{"hasDocs":false,"keywords":["ios","swiftui","xcode"],"lastActivityAt":"2024-01-11T12:05:14Z","packageId":"BF917587-94B7-4E4E-933D-DBAF1AD99AF7","packageName":"AboutKit","packageURL":"\/adamfootdev\/AboutKit","repositoryName":"AboutKit","repositoryOwner":"adamfootdev","stars":98,"summary":"Add an about screen to your app in just a few lines of code."}}}} [component: server]
```

which looks better but is perhaps not ideal to parse.

I think we'll want to drop the non-package results since they don't really take part in normal ordering:

```swift
for result in results.filter(\.isPackage)
```

in line 47. Then we get

```
[ INFO ] GET /search [component: server, request-id: 33710481-DE9C-4A78-81B4-13E467D10CAA]
[ INFO ] search: {"query":"foo","searchID":"6D35A7E2-29E1-4AEC-A4F6-4A224160AFB2"} [component: server]
[ INFO ] searchresult: {"id":"6D35A7E2-29E1-4AEC-A4F6-4A224160AFB2","r":{"package":{"_0":{"hasDocs":false,"keywords":["ios","swiftui","xcode"],"lastActivityAt":"2024-01-11T12:05:14Z","packageId":"BF917587-94B7-4E4E-933D-DBAF1AD99AF7","packageName":"AboutKit","packageURL":"\/adamfootdev\/AboutKit","repositoryName":"AboutKit","repositoryOwner":"adamfootdev","stars":98,"summary":"Add an about screen to your app in just a few lines of code."}}}} [component: server]
```

That's better but might still be too awkward to parse. We could also encode this as CSV:

- searchID, query for the `SearchQuery`
- searchID, index, packageID, + whatever else is needed for the fragments. NB: we can easily query package data via the API, so the packageID itself is enough if we don't want to log lots of detail - but logging the details might be easier if they're needed.

```
[ INFO ] search: 6D35A7E2-29E1-4AEC-A4F6-4A224160AFB2,foo [component: server]
[ INFO ] searchresult: 6D35A7E2-29E1-4AEC-A4F6-4A224160AFB2,0,BF917587-94B7-4E4E-933D-DBAF1AD99AF7 [component: server]
```

I'd imagine this would be easier to parse when we grab the logs. Although the prefix and suffix should be static, so stripping them reliably and then decoding the JSON should also be feasible.

What's definitly missing if we don't use my proposed CSV encoding is the record index. We _can_ probably rely on the log ordering but it'd be safer to embed it in addition:

```swift
        for (idx, result) in results.filter(\.isPackage).enumerated() {
            let fragment = SearchResultFragment(searchID: uniqueSearchID, result: result)
            do {
                let json = String(decoding: try jsonEncoder.encode(fragment), as: UTF8.self)
                AppEnvironment.logger.info("searchresult[\(idx)]: \(json)")
            } catch {
```

yielding:

```
[ INFO ] GET /search [component: server, request-id: 7B7EA929-F2D0-45F6-9430-900DF7E9AB6A]
[ INFO ] search: {"query":"foo","searchID":"2B9FD884-B20B-4319-AB41-08DDBD54A446"} [component: server]
[ INFO ] searchresult[0]: {"id":"2B9FD884-B20B-4319-AB41-08DDBD54A446","r":{"package":{"_0":{"hasDocs":false,"keywords":["ios","swiftui","xcode"],"lastActivityAt":"2024-01-11T12:05:14Z","packageId":"BF917587-94B7-4E4E-933D-DBAF1AD99AF7","packageName":"AboutKit","packageURL":"\/adamfootdev\/AboutKit","repositoryName":"AboutKit","repositoryOwner":"adamfootdev","stars":98,"summary":"Add an about screen to your app in just a few lines of code."}}}} [component: server]
[ INFO ] searchresult[1]: {"id":"2B9FD884-B20B-4319-AB41-08DDBD54A446","r":{"package":{"_0":{"hasDocs":false,"keywords":["parser-combinators","swift","swift-library"],"lastActivityAt":"2019-04-07T21:44:16Z","packageId":"4B10B099-BC77-4687-8B47-ECF0D3A6C8B1","packageName":"FootlessParser","packageURL":"\/kareman\/FootlessParser","repositoryName":"FootlessParser","repositoryOwner":"kareman","stars":63,"summary":"A simple parser combinator written in Swift"}}}} [component: server]
```

I'd say let's try this to our dev environment with the JSON format and test the log-grabbing and decoding process there. Then we can see if it works well enough or if we want it as CSV.

To gate it:

```swift
    static func log(query: String, results: [Search.Result]) {
        guard Current.environment() != .production else { return }
```

I'll see if I can sprinkle these individual snippets in as suggestions as well.

Does that all make sense?

Reply:
Oooh - the index on the result to ensure its position information is a good idea - hopefully it'll be redundant in practice. I dropped the filter part - mostly because I _do_ want to keep the keywords and author bits that are returned. What I want to assemble in the end (for SPISearch) is the equivalent of what the user sees as the search query and totality of the results, to provide all the same contextual clues as to why the result was relevant (or not). That may be overkill, but I think there's value there, especially with evaluation being so essentially subjective.

And yeah, I know it'll be an interesting "joy" to reconstitute from the logs, but I think all the critical bits are there to enable it so that we can walk away with a set/collection of search results (post-processing from the logs) that we can load up into SPISearch, evaluate, share, and run analysis on over time - sort of the big end goal for me.